lib/beacon_web/data_source.ex

```diff
@@ -4,30 +4,20 @@ defmodule BeaconWeb.DataSource do
   require Logger
 
   def live_data(site, path_info, params \\ %{}) when is_atom(site) and is_list(path_info) and is_map(params) do
-    data_source_module = Beacon.Loader.data_source_module_for_site(site)
-
-    if :erlang.module_loaded(data_source_module) do
-      data_source_module.live_data(path_info, params)
-    else
-      Logger.warning("""
-      data source module #{data_source_module} for site #{site} and path #{inspect(path_info)} is not loaded
-
-      returning empty live data for that page
-      """)
-
-      %{}
-    end
+    site
+    |> Beacon.Loader.fetch_live_data_module()
+    |> Beacon.apply_mfa(:live_data, [path_info, params])
   end
 
-  def page_title(assigns) do
+  def page_title(site, page_id, live_data) do
     page =
-      assigns.__dynamic_page_id__
-      |> Beacon.Loader.page_module_for_site()
-      |> Beacon.Loader.call_function_with_retry!(:page_assigns, [])
+      site
+      |> Beacon.Loader.fetch_page_module(page_id)
+      |> Beacon.apply_mfa(:page_assigns, [])
 
-    title = BeaconWeb.Layouts.page_title(assigns)
+    title = Map.get(page, :title) || missing_page_title()
```

Comment:
Can we use `Map.fetch!/2` here instead?  At first glance, it's confusing whether the `nil` is coming from missing `:title` key or if key is present `%{title: nil}`.  But I'm thinking this is a struct so the key should always be present, right?  If so, using `fetch!` seems like it would make that explicit.  WDYT?

Reply:
> Can we use `Map.fetch!/2` here instead?  At first glance, it's confusing whether the `nil` is coming from missing `:title` key or if key is present `%{title: nil}`.  But I'm thinking this is a struct so the key should always be present, right?  If so, using `fetch!` seems like it would make that explicit.  WDYT?

Yep that field should be present so fetch! is the right choice here

Reply:
or maybe `page.title`?